src/dbnode/storage/shard.go

```diff
@@ -955,38 +988,57 @@ func (s *dbShard) tryRetrieveWritableSeries(id ident.ID) (
 
 func (s *dbShard) newShardEntry(
 	id ident.ID,
-	tagsIter ident.TagIterator,
+	tagsArgOpts tagsArgOptions,
 ) (*lookup.Entry, error) {
-	series := s.seriesPool.Get()
 	// NB(r): As documented in storage/series.DatabaseSeries the series IDs
 	// are garbage collected, hence we cast the ID to a BytesID that can't be
 	// finalized.
 	// Since series are purged so infrequently the overhead of not releasing
 	// back an ID to a pool is amortized over a long period of time.
-	clonedID := ident.BytesID(append([]byte(nil), id.Bytes()...))
-	clonedID.NoFinalize()
+	var newID ident.BytesID
+	if id.IsNoFinalize() {
+		// If the ID is already marked as NoFinalize, meaning it won't be returned
+		// to any pools, then we can directly take reference to it.
+		// We make sure to use ident.BytesID for this ID to avoid inc/decref when
+		newID = ident.BytesID(id.Bytes())
+	} else {
+		newID = ident.BytesID(append([]byte(nil), id.Bytes()...))
+		newID.NoFinalize()
+	}
 
 	var (
 		tags ident.Tags
 		err  error
 	)
+	switch tagsArgOpts.arg {
+	case tagsIterArg:
+		tagsIter := tagsArgOpts.tagsIter
 
-	dupTagsIter := tagsIter.Duplicate()
-	tagsIter = nil // Original tagsIter should not be closed
-	if dupTagsIter.Remaining() > 0 {
+		// Ensure tag iterator at start
+		if tagsIter.CurrentIndex() != 0 {
+			return nil, errNewShardEntryTagsIterNotAtIndexZero
+		}
 		tags, err = convert.TagsFromTagsIter(
-			clonedID, dupTagsIter, s.identifierPool)
-		dupTagsIter.Close()
+			newID, tagsIter, s.identifierPool)
+		tagsIter.Close()
 		if err != nil {
 			return nil, err
 		}
 
-		if err := convert.ValidateMetric(clonedID, tags); err != nil {
+		if err := convert.ValidateMetric(newID, tags); err != nil {
 			return nil, err
 		}
+
+	case tagsArg:
+		tags = tagsArgOpts.tags
```

Comment:
this is not pooled / its safe to do this right?

Reply:
Yeah, in the case where we actually pass through an `ident.Tags` it is not pooled and is returned from bootstrappers.